beacon-chain/rpc/prysm/v1alpha1/validator/proposer_bellatrix.go

```diff
@@ -243,6 +248,16 @@ func (vs *Server) getPayloadHeaderFromBuilder(ctx context.Context, slot primitiv
 		return nil, fmt.Errorf("incorrect parent hash %#x != %#x", header.ParentHash(), h.BlockHash())
 	}
 
+	reg, err := vs.BlockBuilder.RegistrationByValidatorID(ctx, idx)
```

Comment:
Does the registration cache  handle the case where the validator does not include a value?

Reply:
I think so. I went through the code as my understanding: the `DefaultBuilderGasLimit` is initially set in `config.go`:  
![Screenshot 2024-12-12 at 10 50 14 AM](https://github.com/user-attachments/assets/74df155d-4fd6-439a-8779-389e7d2a3a4b)

It is then defined as 30M in `mainnet_config.go`:  
![Screenshot 2024-12-12 at 10 49 55 AM](https://github.com/user-attachments/assets/467e6e77-4b3c-41d6-9b71-1bea88acefea)

The beacon node does not provide a flag to set this gas limit, and it is not part of the consensus spec configuration. Instead, it is hard-coded in the client. The `DefaultBuilderGasLimit` is also defined as a default in the validator's `flag.go`. If the validator does not specify the `--suggested-gas-limit` flag, the value in `mainnet_config.go` overrides it:  
![Screenshot 2024-12-12 at 10 52 27 AM](https://github.com/user-attachments/assets/1413efd1-1f32-4db4-9d2d-8d177e819a57)

The `DefaultBuilderGasLimit` is used in `proposer_setting.go`'s `getProposerSettings` function. This function displays or recreates the currently applied proposer settings under the validator's custom flags:  
![Screenshot 2024-12-12 at 10 54 43 AM](https://github.com/user-attachments/assets/5be76796-d96f-4473-9f17-c0f7844b89d4)

In `loader.go`, the `DefaultBuilderGasLimit` ensures that the default gas limit applies if the `--suggested-gas-limit` flag is not set. It is then saved to the proposer settings at startup when registering for the validator node service:  
![Screenshot 2024-12-12 at 11 00 58 AM](https://github.com/user-attachments/assets/1d6ef3c6-c343-429a-92ac-da68f6de7fb9)

The `SaveProposerSettings` function is called within the `Load` method of `settingsLoader` in `loader.go`. This step saves the default gas limit to the proposer settings file:  
![Screenshot 2024-12-12 at 11 04 13 AM](https://github.com/user-attachments/assets/e4fae0dd-cae2-46ab-bd2c-7ad34d21d87a)
![Screenshot 2024-12-12 at 11 04 41 AM](https://github.com/user-attachments/assets/9478699a-5dac-48db-abbb-29b3fc1c0f2d) 

When registering the validator client, the system ensures that the proposer settings with the correct default gas limit are stored in the database:  
![Screenshot 2024-12-12 at 11 05 43 AM](https://github.com/user-attachments/assets/678d2612-772b-4b25-96ba-5bf53509bf41)

Reply:
Not sure I get it right but it seems there is something confusing here, when using `--proposer-settings-file` at the validator level, if in the file if there is:

```
{
  "default_config": {
    "builder": {
      "enabled": false,
      "gas_limit": "42"
    },
    "fee_recipient": "XXXX"
  },
  "my_key": {
    "YYYY": {
      "builder": {
        "enabled": true
       "gas_limit": "43",
      },
      "fee_recipient": "YYYYY"
    },
```

When using `--suggested-fee-recipient=ZZZZ`, my_key will still propose on `YYYY` (proposer settings file has priority over flag).

However, when using `--suggested-gas-limit=10`, my_key will use the flag override instead of `43` (flag has priority for gas limit over proposer settings).

In processBuilderConfig, gas limit is overridden by the flag while the fee recipient is not:

![Screenshot 2024-12-13 at 10 53 22](https://github.com/user-attachments/assets/27d891e9-d4a6-4322-8950-799632797d8a)

Also, if the user specifies for `my_key` a gas limit but no fee-recipient, it seems no fee-recipient is used? (or maybe later in the hot-path it defaults to the flag?).

It could make sense to align both behavior (to always have the file override the flags).

Reply:
yes there's a bug there, similar bug: https://github.com/prysmaticlabs/prysm/issues/14721
unrelated to this PR though so will merge this PR and fix that today